Updated a basic inference for assistant chatbot using Llama3 8B instr…

JannatulFaika opened the pull request:
model using LangChain and transformers.

Updated a basic inference for assistant chatbot using Llama3 8B instruct model using LangChain and transformers.

Using localtunnel to get a public ip. 

to run use this command which is in notebook already
```

!curl ipv4.icanhazip.com
!streamlit run /kaggle/working/app.py &>./logs.txt & npx localtunnel --port 8501
```

JannatulFaika committed:
Updated a basic inference for assistant chatbot using Llama3 8B instruct model using LangChain and transformers.

Updated a basic inference for assistant chatbot using Llama3 8B instruct model using LangChain and transformers.

JannatulFaika committed:
Update requirements.txt

JannatulFaika committed:
Merge pull request #13 from DrAlzahraniProjects/main

Rebasing

JannatulFaika reviewed (commented):
@N-Kel Can you please try hosting the app locally or from kaggle incase there is gpu issues? 
In kaggle use `GPU T4 x 2` and use public ip as tunnel key. I would recommend using kaggle as it's tested hosting from kaggle. 

Also, if you want to host the full model instead of 8bit quantization, just remove the commented lines and change the necessary lines. 

In addition to that, to use TPU we use `bfloat16`. However if we use regular gpu it needs to be changed to `float16`.

PS. the first message reply would take around 500s to complete if hosted from kaggle since it downloads and loads the model. after that first text it takes around 2s to reply. So for demo we can pregame it by sending a Hi beforehand to complete the loading process. Kaggle sometimes stops sessions if there is inactivity so it's better to hover mouse over the window if hosted on kaggle.

N-Kel commented:
On line 40 of app.py, you have an API token hardcoded. That's an [extremely unsafe practice](https://developer.android.com/privacy-and-security/risks/hardcoded-cryptographic-secrets). (E.g. a malicious actor could DoS our app by copying the API key, then running their own app that spams requests so none of ours ever get through. Also, if anything ever happens to your account or the API key--e.g. HuggingFace decides to revoke it--our entire app would die.)
Don't forget that we're going to have to eventually e.g. run our code through Burp Suite and other vulnerability checkers per the professor's instructions.

JannatulFaika commented:
> On line 40 of app.py, you have an API token hardcoded. That's an [extremely unsafe practice](https://developer.android.com/privacy-and-security/risks/hardcoded-cryptographic-secrets). (E.g. a malicious actor could DoS our app by copying the API key, then running their own app that spams requests so none of ours ever get through. Also, if anything ever happens to your account or the API key--e.g. HuggingFace decides to revoke it--our entire app would die.) Don't forget that we're going to have to eventually e.g. run our code through Burp Suite and other vulnerability checkers per the professor's instructions.

Well without the token you can't pull llama from huggingface. also the associated token should have permission to access llama since it's gated model. that's why to avoid last day's OSError the token is provided to avoid any surprises. It's also a fine-grained token which have read access to llama only. 
Also, in the notebook code you can see model caching is used. So in pregame before demo it will download the model and all. Later it will load the model from cache. So, if any issues arise it will be found beforehand. However, huggingface is committed to the token usage as it's the best practice to load gated models from huggingface. However, in final project we can load the keys from environment config file.

N-Kel commented:
> Well without the token you can't pull llama from huggingface.

That is why the user has to create and provide their own token. (There are multiple websites describing why hardcoding credentials *anywhere*, whether in the raw code or an environment config file, is dangerous--e.g. [this](https://blog.gitguardian.com/why-its-urgent-to-deal-with-your-hard-coded-credentials/) which describes how it was the biggest cause of data breaches in 2021, or [this](https://trufflesecurity.com/blog/llms-are-teaching-developers-to-hardcode-api-keys) which describes how despite the risks, LLMs have memorized the bad practice and are perpetuating it. I really do not want to have to explain to the professor why we implemented foul security practices in our app during the presentation if he asks how we implemented it, not do I want to have to explain it to the job recruiters pouring over my GitHub contributions when I try to look for a job.)
What we could do is specify in the instructions that the user has to create a HuggingFace account and create a new token, then implement a field in the app that allows the user to input their token which gets passed to the model. (I did something akin to that for my version.)

> PS. the first message reply would take around 500s to complete if hosted from kaggle since it downloads and loads the model. 

During the demonstration we have to show ourselves cloning the repository and initializing everything from scratch. There won't be any opportunities to preload anything, and we only have five minutes (300s) to do the entire presentation: clone the files, run the build scripts, open the chat model, send a couple of messages, and show Jainam's documentation.

JannatulFaika commented:
> > Well without the token you can't pull llama from huggingface.
> 
> That is why the user has to create and provide their own token. (There are multiple websites describing why hardcoding credentials _anywhere_, whether in the raw code or an environment config file, is dangerous--e.g. [this](https://blog.gitguardian.com/why-its-urgent-to-deal-with-your-hard-coded-credentials/) which describes how it was the biggest cause of data breaches in 2021, or [this](https://trufflesecurity.com/blog/llms-are-teaching-developers-to-hardcode-api-keys) which describes how despite the risks, LLMs have memorized the bad practice and are perpetuating it. I really do not want to have to explain to the professor why we implemented foul security practices in our app during the presentation if he asks how we implemented it, not do I want to have to explain it to the job recruiters pouring over my GitHub contributions when I try to look for a job.) What we could do is specify in the instructions that the user has to create a HuggingFace account and create a new token, then implement a field in the app that allows the user to input their token which gets passed to the model. (I did something akin to that for my version.)
> 
> > PS. the first message reply would take around 500s to complete if hosted from kaggle since it downloads and loads the model.
> 
> During the demonstration we have to show ourselves cloning the repository and initializing everything from scratch. There won't be any opportunities to preload anything, and we only have five minutes (300s) to do the entire presentation: clone the files, run the build scripts, open the chat model, send a couple of messages, and show Jainam's documentation.

I think it's not user who should give the token. Instead the token should be from developer's side. When we chat with MetaAI or any other automated chat it does not really ask for key from user afaik. However, since it's a one time thing to load the model so we can get the token from os.env or config file. Anyways, token is removed and history refreshed as if it was never there to avoid anyone pouring over. 


I understand the 5mins time limit. So we must have to download the model locally or on kaggle. So that it doesn't have to download all the files. Using cache it can load in a bit less time than 500s.  In my understanding there's  not any other option to load 17GB model. I am open to suggestions anyways.

JannatulFaika committed:
Update to incorporate huggingface token and quantized model

JannatulFaika committed:
Update requirements.txt

JannatulFaika committed:
Add notebook to deploy from kaggle using localtunnel

use GPU T4 x 2 and use ip as password for tunnel

JannatulFaika committed:
Update getenv for token

JannatulFaika reviewed (commented):
To run the streamlit modify the command starting like this

`HUGGINGFACE_TOKEN=token streamlit run ........Rest of the stuff ..... `

JannatulFaika committed:
Update with streaming

JannatulFaika committed:
Update Dockerfile
